### PR TITLE
Run replicas without the master

### PR DESCRIPTION
**Pull request by alexeyklyukin**
Support creation of replicas in the cluster without the master if one of the replica creation methods allows this (i.e. by getting data from FTP or S3).
It's based on the code that allows the replicafrom tag, so this branch also includes the code from https://github.com/zalando/patroni/pull/113

### CODE REVIEW

File: patroni/ha.py [hunk 1]
```diff
@@ -62,12 +63,12 @@ def touch_member(self):
         self.dcs.touch_member(json.dumps(data, separators=(',', ':')))
 
     def copy_backup_from_leader(self, leader):
-        if self.state_handler.bootstrap(leader):
-            logger.info('bootstrapped from leader')
+        if self.state_handler.bootstrap(True, leader):
```
Comment (a1exsh) [inline patroni/ha.py]: `(initialize=True, current_leader=leader)` please :-)

**Comment by feikesteenbergen** [comment]
+1

**Comment by CyberDem0n** [comment]
Looks good to me

File: patroni/dcs.py [hunk 2]
```diff
@@ -107,6 +111,9 @@ class Cluster(namedtuple('Cluster', 'initialize,leader,last_leader_operation,mem
     def is_unlocked(self):
         return not (self.leader and self.leader.name)
 
+    def has_member(self, member_name):
+        return len([m for m in self.members if m.name == member_name]) > 0
```
Comment (a1exsh) [inline patroni/dcs.py]: `any()`?

File: patroni/ha.py [hunk 3]
```diff
@@ -62,12 +63,12 @@ def touch_member(self):
         self.dcs.touch_member(json.dumps(data, separators=(',', ':')))
 
     def copy_backup_from_leader(self, leader):
```
Comment (a1exsh) [inline patroni/ha.py]: `initialize_replica(self, leader=None)`?

File: patroni/postgresql.py [hunk 4]
```diff
@@ -678,7 +713,7 @@ def bootstrap(self, current_leader=None):
         else:
             if self.sync_from_leader(current_leader):
                 self.restore_configuration_files()
-                self.write_recovery_conf(current_leader)
+                self.write_recovery_conf(current_leader, True)
```
Comment (a1exsh) [inline patroni/postgresql.py]: `, bootstrap=True)`

File: patroni/ha.py [hunk 5]
```diff
@@ -91,41 +92,38 @@ def bootstrap(self):
...
+        ret = demote_reason if (not recovery and self.state_handler.is_leader() or
+                                recovery and self.state_handler.role == 'master') else follow_reason
+        # determine the node to follow. If replicatefrom tag is set,
+        # try to follow the node mentioned there, otherwise, follow the leader.
+        if self.patroni.replicatefrom:
+            node_to_follow = [m for m in self.cluster.members if m.name == self.patroni.replicatefrom]
+            node_to_follow = node_to_follow[0] if node_to_follow else self.cluster.leader
+        else:
+            node_to_follow = self.cluster.leader
+        node_to_follow = None if (node_to_follow and node_to_follow.name) == self.state_handler.name else node_to_follow
```
Comment (a1exsh) [inline patroni/ha.py]: `node_to_follow and node_to_follow.name == self.state_handler_name` please!

File: patroni/postgresql.py [hunk 6]
```diff
@@ -220,9 +220,9 @@ def write_pgpass(self, record):
         return env
 
     def sync_from_leader(self, leader):
```
Comment (a1exsh) [inline patroni/postgresql.py]: Rename function due to leader allowed to be omitted?

File: patroni/postgresql.py [hunk 7]
```diff
@@ -637,7 +651,18 @@ def sync_replication_slots(self, cluster):
...
                 self.load_replication_slots()
-                slots = [m.name for m in cluster.members if m.name != self.name] if self.role == 'master' else []
+                # if the replicatefrom tag is set on the member - we should not create the replication slot for it on
+                # the current master, because that member would replicate from elsewhere. We still create the slot if
+                # the replicatefrom destination member is currently not a member of the cluster (fallback to the
+                # master), or if replicatefrom destination member happens to be the current master
+                if self.role == 'master':
+                    slots = [m.name for m in cluster.members if m.name != self.name and
+                             (not cluster.has_member(m.replicatefrom)
+                              if m.replicatefrom and m.replicatefrom != self.name else True)]
```
Comment (a1exsh) [inline patroni/postgresql.py]: `m.replicatefrom is None or m.replicatefrom == self.name or not cluster.has_member(m.replicatefrom)`

**Pull request merged**